### PR TITLE
[pigeon] adds option for java output to have the Generated annotation

### PR DESCRIPTION
**Pull request by gaaclarke**
fixes https://github.com/flutter/flutter/issues/105653

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by zhenqiu1101** [comment]
LGTM. But I can't find the approval button on my UI. I may not have enough permission.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/pigeon/lib/java_generator.dart [hunk 1]
```diff
@@ -25,6 +26,11 @@ class JavaOptions {
   /// A copyright header that will get prepended to generated code.
   final Iterable<String>? copyrightHeader;
 
+  /// Determines if the `javax.annotation.Generated` is used in the output. This
+  /// is false by default since that dependency isn't available in plugins by
+  /// default .
+  final bool? useGeneratedAnnotation;
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply packages/pigeon/lib/java_generator.dart:32]: Null means the option hasn't been specified, thus use the default value.  It was originally implemented as nonnull but it resulted in duplication of the default value.  By keeping it undefined until it's actually used I have a single point of truth about what the default value is.

File: packages/pigeon/test/pigeon_lib_test.dart [hunk 2]
```diff
@@ -98,6 +98,12 @@ void main() {
     expect(opts.cppHeaderOut, equals('foo.h'));
   });
 
+  test('parse args - java_use_generated_annotation', () {
+    final PigeonOptions opts =
+        Pigeon.parseArgs(<String>['--java_use_generated_annotation']);
+    expect(opts.javaOptions!.useGeneratedAnnotation, isTrue);
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply packages/pigeon/test/pigeon_lib_test.dart:104]: Done

**Review by stuartmorgan — APPROVED** [review]
LGTM

**Pull request merged**